### ISSUE TITLE
Yup validation transforms date string to datetime

### ISSUE THREAD

**Issue by chrisbag** [react-hook-form/react-hook-form#7730]
### Version Number

V7

### Codesandbox/Expo snack

https://codesandbox.io/s/react-hook-form-v7-yup-and-dates-p80hu

### Steps to reproduce

Hello,
First of all thanks for this wonderful library and the new V7.
I have been experiencing an unexpected behavior when using react-hook-form with yup for date validation. This also happens when using v6 I believe.

1. Go to sandbox
2. Choose a date
3. Click submit and check console
4. Date field value in watch is a string of type yyyy-mm-dd
5. However data from on submit is of type Date (basically a timestamp)

### Expected behaviour

I would expect form data not not to be transformed by yup resolver, meaning formState = data passed in on submit. I assume what's happening is that Yup parse the date string into a Date Object and that is the element returned

### Relevant log output

```
submited data =  Thu Feb 03 2022 00:00:00 GMT+0100 (heure normale d’Europe centrale)
form Data =  2022-02-03 
```

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by bluebill1049** [comment]
native input doesn't return `date` type instead of `string`.

**Comment by Moshyfawn** [comment]
Huh, if you, say,[ use zod with it's RHF resolver](https://codesandbox.io/s/react-hook-form-v7-yup-and-dates-forked-nhd1h?file=/src/index.js), you'd get `Expected date, received string` on submit validation. But in the OP's case, yup silently transforms the date string into date and successfully submits it. I'd think, yup should also report an error instead of transforming the value according to the set type

**Comment by bluebill1049** [comment]
oh, I miss the part of the data transform in the `resolver`, we just return what's returned by the schema validation library. we don't need any parsing/transform on the data.

**Comment by Moshyfawn** [comment]
So, the question is then": "does Yup parse the value and return it parsed?"

**Comment by bluebill1049** [comment]
> So, the question is then": "does Yup parse the value and return it parsed?"

I think it doesn't then from the example showed above.

**Comment by bluebill1049** [comment]
I just did a quick reading, i think in `yup`, you will need to use `.transform(parseDateString)` API

**Comment by Moshyfawn** [comment]
If you return a non-date type in `.transform()`, Yup will yell the "Incorrect type" error instead. See [this CSB](https://codesandbox.io/s/react-hook-form-v7-yup-and-dates-forked-ljr6z?file=/src/index.js)

**Comment by bluebill1049** [comment]
If that's the case, I think it will have to combine with `setValueAs` for uncontrolled input or `setState` with the transform with controlled inputs.

**Comment by Moshyfawn** [comment]
IMO, the date type in the OP's schema should be
- **either** `string` with some date validation in `transform` or wherever they do it with Yup **if** OP needs the string
- **or** `date` combined with `valueAsDate: true` as an option to `register`.

I'd say, to keep the value as date would be easier to validate and you can always parse it `onSubmit` into whatever form you need

**Comment by Moshyfawn** [comment]
Basically, the original OP's schema should fail anyway as the type of value sent to the validation schema is `string` which is incorrect with the expected type in validation schema set as `date` :P

**Comment by chrisbag** [comment]
> oh, I miss the part of the data transform in the `resolver`, we just return what's returned by the schema validation library. we don't do any parsing/transform on the data.
> 
> 1. hook form handle user behavior
> 2. validation is done by schema
> 3. data return by the schema validation library https://github.com/react-hook-form/resolvers/blob/master/yup/src/yup.ts#L62

Hey, Thanks a a lot to you both for your reactivity.
Ok This is what I thought, yup parses the data.
Should we really expect to get the data returned by the validation schema ? Obviously we should get the errors but was really surprised to see the data transformed by the validator. In my view, the validation library should not transform the data, it creates a blackbox.
What do you think ?

> At the end of the day, the original OP's schema should fail as the type of value sent to the validation schema is string which is incorrect with the expected type in validation schema set as date :P

Yup casts input value into the correct type, which just makes from validation a lot easier as most input returned is string. have to manually transform each field to use yup would almost defeat the purpose of a validation library

**Comment by bluebill1049** [comment]
Thanks for your input, `resolver` is a standard core lib that we maintaining not just for Yup, as you would see there is a list of resolvers for different schema libraries. I wouldn't consider making a change at the moment as it impacts all of them. Especially this should really come from the schema validation library instead of hook form.

**Comment by Moshyfawn** [comment]
>  I wouldn't consider making a change at the moment

What change are we talking about? Yup transforms the values "the help with validation" which is very subjective and a totally different form, say, Zod's paradigm, where they sure allow you to preprocess the value before validating, but they don't mutate the actual value. It's not like it's RHF job to bring everyone "up to standard", as it all depends on the validation lib ppl choose the lib for they way they hand the data _(transform the data or not in case of Yup and Zod)_. If RHF takes the responsibility to update _(or not)_ the validated data, there gonna be tons of ppl with different libs creating issues on how the behavior of RHF doesn't match their lib.

To sum it up, IMO, if you use Yup, you can expect the values to "mutated" during the validation, as Yup does exactly that. If you use, say, Zod, it's their paradigm that you chose and don't expect the values to be mutated during the validation

**Comment by chrisbag** [comment]
Thanks for your views that I completely understand and it makes senses. I hope this will help out some react-hook-form/ yup users stumbling into the same issue. I was surprised that one would potentially expect the validation library to mutate data. 
I'll probably just make my custom resolver to solve this issue 😁
thanks for your time

**Comment by bluebill1049** [comment]
np, thank you @Moshyfawn's detailed suggestion and help as well.
